// @progress/kendo-angular-progressbar/dist/npm/common/progressbar-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("./util");
/**
 * @hidden
 */
var ProgressBarBase = /** @class */ (function () {
    /**
     * @hidden
     */
    function ProgressBarBase(localization) {
        var _this = this;
        this.localization = localization;
        this.widgetClasses = true;
        this.roleAttribute = 'progressbar';
        /**
         * The maximum value of the ProgressBar.
         * Defaults to `100`.
         */
        this.max = 100;
        /**
         * The minimum value of the ProgressBar.
         * Defaults to `0`.
         */
        this.min = 0;
        /**
         * The value of the ProgressBar.
         * Has to be between `min` and `max`.
         *
         * By default, the value will be equal to the `min` value.
         */
        /**
         * The value of the ProgressBar.
         * Has to be between `min` and `max`.
         *
         * By default, the value will be equal to the 0.
         */
        this.value = 0;
        /**
         * Defines the orientation of the ProgressBar
         * ([see example]({% slug progressbar_orientation %})).
         * Defaults to `horizontal`.
         */
        this.orientation = 'horizontal';
        /**
         * If set to `true`, the ProgressBar will be disabled
         * ([see example]({% slug progressbar_disabled %})).
         * It will still allow you to change its value.
         * Defaults to `false`.
         */
        this.disabled = false;
        /**
         * If set to `true`, the ProgressBar will be reversed
         * ([see example]({% slug progressbar_direction %})).
         * Defaults to `false`.
         */
        this.reverse = false;
        /**
         * Sets the `indeterminate` state of the ProgressBar.
         * Defaults to `false`.
         */
        this.indeterminate = false;
        this.displayValue = 0;
        this.previousValue = 0;
        this.localizationChangeSubscription = localization.changes.subscribe(function (_a) {
            var rtl = _a.rtl;
            _this.direction = rtl ? 'rtl' : 'ltr';
        });
    }
    Object.defineProperty(ProgressBarBase.prototype, "isHorizontal", {
        get: function () {
            return this.orientation === 'horizontal';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "isVertical", {
        get: function () {
            return this.orientation === 'vertical';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "disabledClass", {
        get: function () {
            return this.disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "reverseClass", {
        get: function () {
            return this.reverse;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "indeterminateClass", {
        get: function () {
            return this.indeterminate;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "dirAttribute", {
        get: function () {
            return this.direction;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "ariaMinAttribute", {
        get: function () {
            return String(this.min);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "ariaMaxAttribute", {
        get: function () {
            return String(this.max);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "ariaValueAttribute", {
        get: function () {
            return this.indeterminate ? undefined : String(this.displayValue);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "isCompleted", {
        /**
         * @hidden
         */
        get: function () {
            return this.value === this.max;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusWidth", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'horizontal' ? this._progressRatio * 100 : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusHeight", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'vertical' ? this._progressRatio * 100 : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusWrapperWidth", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'horizontal' ? 100 / this._progressRatio : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusWrapperHeight", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'vertical' ? 100 / this._progressRatio : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "_progressRatio", {
        get: function () {
            return util_1.calculateRatio(this.min, this.max, this.displayValue);
        },
        enumerable: true,
        configurable: true
    });
    ProgressBarBase.prototype.ngOnChanges = function (changes) {
        var min = util_1.extractValueFromChanges(changes, 'min', this.min);
        var max = util_1.extractValueFromChanges(changes, 'max', this.max);
        var value = util_1.extractValueFromChanges(changes, 'value', this.value);
        if (changes.min || changes.max || changes.value) {
            if (changes.min || changes.max) {
                util_1.validateRange(min, max);
            }
            if (changes.value) {
                if (value == null || Number.isNaN(value)) {
                    this.value = min;
                }
                var previousValue = this.displayValue;
                this.displayValue = util_1.adjustValueToRange(this.min, this.max, value);
                if (previousValue !== this.displayValue) {
                    this.previousValue = previousValue;
                }
            }
            this.min = min;
            this.max = max;
            this.displayValue = util_1.adjustValueToRange(this.min, this.max, value);
        }
    };
    ProgressBarBase.prototype.ngOnDestroy = function () {
        if (this.localizationChangeSubscription) {
            this.localizationChangeSubscription.unsubscribe();
        }
    };
    ProgressBarBase.propDecorators = {
        widgetClasses: [{ type: core_1.HostBinding, args: ['class.k-widget',] }, { type: core_1.HostBinding, args: ['class.k-progressbar',] }],
        isHorizontal: [{ type: core_1.HostBinding, args: ['class.k-progressbar-horizontal',] }],
        isVertical: [{ type: core_1.HostBinding, args: ['class.k-progressbar-vertical',] }],
        disabledClass: [{ type: core_1.HostBinding, args: ['class.k-state-disabled',] }],
        reverseClass: [{ type: core_1.HostBinding, args: ['class.k-progressbar-reverse',] }],
        indeterminateClass: [{ type: core_1.HostBinding, args: ['class.k-progressbar-indeterminate',] }],
        dirAttribute: [{ type: core_1.HostBinding, args: ['attr.dir',] }],
        roleAttribute: [{ type: core_1.HostBinding, args: ['attr.role',] }],
        ariaMinAttribute: [{ type: core_1.HostBinding, args: ['attr.aria-valuemin',] }],
        ariaMaxAttribute: [{ type: core_1.HostBinding, args: ['attr.aria-valuemax',] }],
        ariaValueAttribute: [{ type: core_1.HostBinding, args: ['attr.aria-valuenow',] }],
        max: [{ type: core_1.Input }],
        min: [{ type: core_1.Input }],
        value: [{ type: core_1.Input }],
        orientation: [{ type: core_1.Input }],
        disabled: [{ type: core_1.Input }],
        reverse: [{ type: core_1.Input }],
        indeterminate: [{ type: core_1.Input }]
    };
    return ProgressBarBase;
}());
exports.ProgressBarBase = ProgressBarBase;
