// @progress/kendo-angular-toolbar/dist/es2015/tools/toolbar-tool.component.js
import { Component, Input } from '@angular/core';
import { ToolNavigationService } from '../navigation/tool-navigation.service';
/**
 * Represents the Base ToolBar Tool component for Angular.
 * Extend this class to create custom tools.
 */
export class ToolBarToolComponent {
    constructor() {
        this.tabIndex = -1; //Focus movement inside the toolbar is managed using roving tabindex.
        this.overflows = true;
        // this should be replaced with showTool: DisplayMode = 'both';
        /**
         * @hidden
         */
        this.responsive = true;
        if (!this.navigationService) {
            this.navigationService = new ToolNavigationService();
        }
    }
    get toolbarDisplay() {
        return this.overflows ? 'none' : 'inline-block';
    }
    get overflowDisplay() {
        return this.overflows ? 'block' : 'none';
    }
}
ToolBarToolComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: 'toolbar-tool',
                template: ``
            },] },
];
/** @nocollapse */
ToolBarToolComponent.ctorParameters = () => [];
ToolBarToolComponent.propDecorators = {
    responsive: [{ type: Input }]
};
