// @progress/kendo-angular-layout/dist/es2015/common/dom-queries.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2019 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ITEM_INDEX } from '../drawer/models/constants';
import { isPresent } from '../drawer/util';
const focusableRegex = /^(?:a|input|select|option|textarea|button|object)$/i;
const toClassList = (classNames) => String(classNames).trim().split(' ');
const ɵ0 = toClassList;
/**
 * @hidden
 */
export const isFocusable = (element) => {
    if (element.tagName) {
        const tagName = element.tagName.toLowerCase();
        const tabIndex = element.getAttribute('tabIndex');
        const skipTab = tabIndex === '-1';
        let focusable = tabIndex !== null && !skipTab;
        if (focusableRegex.test(tagName)) {
            focusable = !element.disabled && !skipTab;
        }
        return focusable;
    }
    return false;
};
/**
 * @hidden
 */
export const hasClass = (element, className) => Boolean(toClassList(element.className).find((name) => name === className));
/**
 * @hidden
 */
export const closestInScope = (target, predicate, scope) => {
    while (target && target !== scope && !predicate(target)) {
        target = target.parentNode;
    }
    if (target !== scope) {
        return target;
    }
};
/**
 * @hidden
 */
export const itemIndex = (item) => +item.getAttribute(ITEM_INDEX);
/**
 *
 * @hidden
 */
export const hasItemIndex = (item) => isPresent(item.getAttribute(ITEM_INDEX));
/**
 * @hidden
 */
export const closestItem = (target, scope) => closestInScope(target, hasItemIndex, scope);
export { ɵ0 };
