// @progress/kendo-angular-layout/dist/es/common/dom-queries.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2019 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ITEM_INDEX } from '../drawer/models/constants';
import { isPresent } from '../drawer/util';
var focusableRegex = /^(?:a|input|select|option|textarea|button|object)$/i;
var toClassList = function (classNames) { return String(classNames).trim().split(' '); };
var ɵ0 = toClassList;
/**
 * @hidden
 */
export var isFocusable = function (element) {
    if (element.tagName) {
        var tagName = element.tagName.toLowerCase();
        var tabIndex = element.getAttribute('tabIndex');
        var skipTab = tabIndex === '-1';
        var focusable = tabIndex !== null && !skipTab;
        if (focusableRegex.test(tagName)) {
            focusable = !element.disabled && !skipTab;
        }
        return focusable;
    }
    return false;
};
/**
 * @hidden
 */
export var hasClass = function (element, className) {
    return Boolean(toClassList(element.className).find(function (name) { return name === className; }));
};
/**
 * @hidden
 */
export var closestInScope = function (target, predicate, scope) {
    while (target && target !== scope && !predicate(target)) {
        target = target.parentNode;
    }
    if (target !== scope) {
        return target;
    }
};
/**
 * @hidden
 */
export var itemIndex = function (item) { return +item.getAttribute(ITEM_INDEX); };
/**
 *
 * @hidden
 */
export var hasItemIndex = function (item) { return isPresent(item.getAttribute(ITEM_INDEX)); };
/**
 * @hidden
 */
export var closestItem = function (target, scope) { return closestInScope(target, hasItemIndex, scope); };
export { ɵ0 };
