// @progress/kendo-angular-layout/dist/npm/common/dom-queries.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2019 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("../drawer/models/constants");
var util_1 = require("../drawer/util");
var focusableRegex = /^(?:a|input|select|option|textarea|button|object)$/i;
var toClassList = function (classNames) { return String(classNames).trim().split(' '); };
var ɵ0 = toClassList;
exports.ɵ0 = ɵ0;
/**
 * @hidden
 */
exports.isFocusable = function (element) {
    if (element.tagName) {
        var tagName = element.tagName.toLowerCase();
        var tabIndex = element.getAttribute('tabIndex');
        var skipTab = tabIndex === '-1';
        var focusable = tabIndex !== null && !skipTab;
        if (focusableRegex.test(tagName)) {
            focusable = !element.disabled && !skipTab;
        }
        return focusable;
    }
    return false;
};
/**
 * @hidden
 */
exports.hasClass = function (element, className) {
    return Boolean(toClassList(element.className).find(function (name) { return name === className; }));
};
/**
 * @hidden
 */
exports.closestInScope = function (target, predicate, scope) {
    while (target && target !== scope && !predicate(target)) {
        target = target.parentNode;
    }
    if (target !== scope) {
        return target;
    }
};
/**
 * @hidden
 */
exports.itemIndex = function (item) { return +item.getAttribute(constants_1.ITEM_INDEX); };
/**
 *
 * @hidden
 */
exports.hasItemIndex = function (item) { return util_1.isPresent(item.getAttribute(constants_1.ITEM_INDEX)); };
/**
 * @hidden
 */
exports.closestItem = function (target, scope) { return exports.closestInScope(target, exports.hasItemIndex, scope); };
