// @progress/kendo-angular-scheduler/dist/es2015/events/view-event-map.js
import { CreateEvent } from './create-event';
import { DragEndEvent } from './drag-end-event';
import { DragEvent } from './drag-event';
import { DragStartEvent } from './drag-start-event';
import { EventClickEvent } from './event-click-event';
import { EventKeydownEvent } from './event-keydown-event';
import { RemoveEvent } from './remove-event';
import { ResizeEndEvent } from './resize-end-event';
import { ResizeEvent } from './resize-event';
import { ResizeStartEvent } from './resize-start-event';
import { SlotClickEvent } from './slot-click-event';
/**
 * @hidden
 *
 * Maps the name of the event to the argument type of the event.
 */
export const VIEW_EVENT_MAP = {
    slotClick: SlotClickEvent,
    slotDblClick: SlotClickEvent,
    eventClick: EventClickEvent,
    eventDblClick: EventClickEvent,
    eventKeydown: EventKeydownEvent,
    create: CreateEvent,
    remove: RemoveEvent,
    resizeStart: ResizeStartEvent,
    resize: ResizeEvent,
    resizeEnd: ResizeEndEvent,
    dragStart: DragStartEvent,
    drag: DragEvent,
    dragEnd: DragEndEvent
};
