// @progress/kendo-angular-scheduler/dist/npm/navigation/shortcuts.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var rxjs_1 = require("rxjs");
var focus_service_1 = require("./focus.service");
var modifiers_1 = require("../common/modifiers");
var util_1 = require("../common/util");
var scheduler_component_1 = require("../scheduler.component");
var dom_events_service_1 = require("../views/common/dom-events.service");
var operators_1 = require("rxjs/operators");
var view_state_service_1 = require("../views/view-state.service");
var dom_queries_1 = require("../common/dom-queries");
var dialogs_service_1 = require("../editing/dialogs.service");
var isContentWrapper = function (element) { return dom_queries_1.hasClasses(element, 'k-scheduler-content'); };
var ɵ0 = isContentWrapper;
exports.ɵ0 = ɵ0;
/**
 * @hidden
 */
var ShortcutsDirective = /** @class */ (function () {
    function ShortcutsDirective(scheduler, domEvents, focusService, zone, changeDetector, viewState, dialogsService) {
        var _this = this;
        this.scheduler = scheduler;
        this.domEvents = domEvents;
        this.focusService = focusService;
        this.zone = zone;
        this.changeDetector = changeDetector;
        this.viewState = viewState;
        this.dialogsService = dialogsService;
        this.shortcuts = [{
                match: function (e) { return e.keyCode === kendo_angular_common_1.Keys.KeyC && modifiers_1.noModifiers(e); },
                action: function (e) {
                    var scheduler = _this.scheduler;
                    var hours = new Date().getHours();
                    var selected = scheduler.selectedDate;
                    var start = new Date(selected.getFullYear(), selected.getMonth(), selected.getDate(), hours + 1);
                    var end = new Date(selected.getFullYear(), selected.getMonth(), selected.getDate(), hours + 2);
                    var firstResource;
                    if (scheduler.group) {
                        var resources = scheduler.resources || [];
                        var group = scheduler.group || {};
                        var grouped = util_1.groupResources(group, resources);
                        if (grouped.length > 0) {
                            firstResource = grouped[0].data[0];
                        }
                    }
                    scheduler.create.emit({
                        start: kendo_date_math_1.ZonedDate.fromLocalDate(start, scheduler.timezone).toLocalDate(),
                        end: kendo_date_math_1.ZonedDate.fromLocalDate(end, scheduler.timezone).toLocalDate(),
                        isAllDay: false,
                        resources: [firstResource],
                        originalEvent: e,
                        sender: scheduler
                    });
                }
            }, {
                match: function (e) { return e.keyCode >= kendo_angular_common_1.Keys.Digit1 && e.keyCode <= kendo_angular_common_1.Keys.Digit9 && modifiers_1.withModifiers(e, modifiers_1.Modifiers.AltKey); },
                action: function (e) {
                    var scheduler = _this.scheduler;
                    var viewIndex = e.keyCode - kendo_angular_common_1.Keys.Digit0 - 1;
                    var views = scheduler.views.toArray();
                    var view = views[viewIndex];
                    if (view) {
                        _this.zone.run(function () {
                            var prevented = scheduler.onNavigationAction({ type: 'view-change', view: view });
                            if (!prevented) {
                                _this.changeDetector.markForCheck();
                                _this.focusWait();
                            }
                        });
                    }
                }
            }, {
                match: function (e) { return e.keyCode === kendo_angular_common_1.Keys.KeyT && modifiers_1.noModifiers(e); },
                action: function () {
                    _this.zone.run(function () {
                        _this.scheduler.onNavigationAction({ type: 'today' });
                        _this.focusWait();
                    });
                }
            }, {
                // tslint:disable-next-line:deprecation
                match: function (e) { return (e.keyCode === kendo_angular_common_1.Keys.ArrowLeft || e.keyCode === kendo_angular_common_1.Keys.ArrowRight) && modifiers_1.withModifiers(e, modifiers_1.Modifiers.ShiftKey); },
                action: function (e) {
                    // tslint:disable-next-line:deprecation
                    var type = e.keyCode === kendo_angular_common_1.Keys.ArrowLeft ? 'prev' : 'next';
                    _this.zone.run(function () {
                        _this.scheduler.onNavigationAction({ type: type });
                        _this.focusWait();
                    });
                }
            }, {
                // tslint:disable-next-line:deprecation
                match: function (e) { return (e.keyCode === kendo_angular_common_1.Keys.ArrowUp || e.keyCode === kendo_angular_common_1.Keys.ArrowLeft) && modifiers_1.noModifiers(e) && !isContentWrapper(e.target); },
                action: function (e) {
                    var prevented = _this.scheduler.onNavigationAction({ type: 'focus-prev' });
                    if (!prevented) {
                        _this.focusService.focusNext({ offset: -1 });
                        e.preventDefault();
                    }
                }
            }, {
                // tslint:disable-next-line:deprecation
                match: function (e) { return (e.keyCode === kendo_angular_common_1.Keys.ArrowDown || e.keyCode === kendo_angular_common_1.Keys.ArrowRight) && modifiers_1.noModifiers(e) && !isContentWrapper(e.target); },
                action: function (e) {
                    var prevented = _this.scheduler.onNavigationAction({ type: 'focus-next' });
                    if (!prevented) {
                        _this.focusService.focusNext({ offset: 1 });
                        e.preventDefault();
                    }
                }
            }];
        this.taskShortcuts = [{
                match: function (e) { return (e.keyCode === kendo_angular_common_1.Keys.Delete || e.keyCode === kendo_angular_common_1.Keys.Backspace) && modifiers_1.noModifiers(e); },
                action: function (e, event) {
                    _this.viewState.emitEvent('remove', { event: event, dataItem: event.dataItem });
                    e.preventDefault();
                }
            }, {
                match: function (e) { return e.keyCode === kendo_angular_common_1.Keys.Enter && modifiers_1.noModifiers(e); },
                action: function (e, event) {
                    _this.viewState.emitEvent('eventDblClick', { type: 'dblclick', event: event, originalEvent: e });
                    e.preventDefault();
                }
            }];
        this.subs = new rxjs_1.Subscription();
        this.subs.add(this.domEvents.keydown.subscribe(function (e) { return _this.onKeydown(e); }));
        this.subs.add(this.scheduler.eventKeydown.subscribe(function (e) { return _this.onEventKeydown(e); }));
    }
    ShortcutsDirective.prototype.onKeydown = function (e) {
        var match = this.shortcuts.find(function (shortcut) { return shortcut.match(e); });
        if (match && !this.dialogsService.isOpen) {
            match.action(e);
        }
    };
    ShortcutsDirective.prototype.onEventKeydown = function (e) {
        var match = this.taskShortcuts.find(function (shortcut) { return shortcut.match(e.originalEvent); });
        if (match && !this.dialogsService.isOpen) {
            match.action(e.originalEvent, e.event);
        }
    };
    ShortcutsDirective.prototype.focusWait = function () {
        var _this = this;
        this.viewState.layoutEnd.pipe(operators_1.take(1)).subscribe(function () {
            return _this.focusService.focus();
        });
    };
    ShortcutsDirective.decorators = [
        { type: core_1.Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'kendo-scheduler'
                },] },
    ];
    /** @nocollapse */
    ShortcutsDirective.ctorParameters = function () { return [
        { type: scheduler_component_1.SchedulerComponent },
        { type: dom_events_service_1.DomEventsService },
        { type: focus_service_1.FocusService },
        { type: core_1.NgZone },
        { type: core_1.ChangeDetectorRef },
        { type: view_state_service_1.ViewStateService },
        { type: dialogs_service_1.DialogsService }
    ]; };
    return ShortcutsDirective;
}());
exports.ShortcutsDirective = ShortcutsDirective;
