// @progress/kendo-angular-progressbar/dist/fesm5/index.js
import { __extends } from 'tslib';
import { Component, EventEmitter, HostBinding, Input, NgModule, NgZone, Output, Renderer2, ViewChild, isDevMode } from '@angular/core';
import { L10N_PREFIX, LocalizationService } from '@progress/kendo-angular-l10n';
import { hasObservers } from '@progress/kendo-angular-common';
import { CommonModule } from '@angular/common';

/**
 * @hidden
 */
var MIN_MAX_ERROR_MESSAGE = "The max value should be greater than the min.";
/**
 * @hidden
 */
var LABEL_DECIMALS = 3;
/**
 * @hidden
 */
var MIN_RATIO = 0.00001;

/**
 * @hidden
 */
var reverseChunks = function (orientation, reverse) { return (orientation === 'vertical' && !reverse || orientation === 'horizontal' && reverse); };
/**
 * @hidden
 */
var formatValue = function (value, min, max, label) {
    var defaultFormattedValue = truncateNumber(value);
    if (typeof label !== 'boolean') {
        if (typeof label.format === 'string') {
            switch (label.format) {
                case 'value':
                    return defaultFormattedValue;
                case 'percent':
                    return Math.floor(calculatePercentage(value, min, max)) + "%";
                default:
                    return defaultFormattedValue;
            }
        }
        else if (typeof label.format === 'function') {
            return label.format(value);
        }
        else {
            return defaultFormattedValue;
        }
    }
    return defaultFormattedValue;
};
/**
 * @hidden
 */
var validateRange = function (min, max) {
    if (isDevMode && min > max) {
        throw new Error(MIN_MAX_ERROR_MESSAGE);
    }
};
/**
 * @hidden
 */
var adjustValueToRange = function (min, max, value) { return Math.max(Math.min(value, max), min); };
/**
 * @hidden
 */
var calculatePercentage = function (value, min, max) {
    var onePercent = Math.abs((max - min) / 100);
    return Math.abs((value - min) / onePercent);
};
/**
 * @hidden
 */
var truncateNumber = function (value) {
    var numberParts = value.toString().split('.');
    return numberParts.length === 1 ? "" + numberParts[0] : numberParts[0] + "." + numberParts[1].substr(0, LABEL_DECIMALS);
};
/**
 * @hidden
 */
var calculateRatio = function (min, max, value) { return Math.max((value - min) / (max - min), MIN_RATIO); };
/**
 * @hidden
 */
var extractValueFromChanges = function (changes, type, value) {
    return changes[type] && changes[type].currentValue !== undefined ? changes[type].currentValue : value;
};
/**
 * @hidden
 */
var runAnimation = function (changes, animation, previousValue, displayValue) { return animation && changes.value && previousValue !== displayValue; };

/**
 * @hidden
 */
var ProgressBarBase = /** @class */ (function () {
    /**
     * @hidden
     */
    function ProgressBarBase(localization) {
        var _this = this;
        this.localization = localization;
        this.widgetClasses = true;
        this.roleAttribute = 'progressbar';
        /**
         * The maximum value of the ProgressBar.
         * Defaults to `100`.
         */
        this.max = 100;
        /**
         * The minimum value of the ProgressBar.
         * Defaults to `0`.
         */
        this.min = 0;
        /**
         * The value of the ProgressBar.
         * Has to be between `min` and `max`.
         *
         * By default, the value will be equal to the `min` value.
         */
        /**
         * The value of the ProgressBar.
         * Has to be between `min` and `max`.
         *
         * By default, the value will be equal to the 0.
         */
        this.value = 0;
        /**
         * Defines the orientation of the ProgressBar
         * ([see example]({% slug progressbar_orientation %})).
         * Defaults to `horizontal`.
         */
        this.orientation = 'horizontal';
        /**
         * If set to `true`, the ProgressBar will be disabled
         * ([see example]({% slug progressbar_disabled %})).
         * It will still allow you to change its value.
         * Defaults to `false`.
         */
        this.disabled = false;
        /**
         * If set to `true`, the ProgressBar will be reversed
         * ([see example]({% slug progressbar_direction %})).
         * Defaults to `false`.
         */
        this.reverse = false;
        /**
         * Sets the `indeterminate` state of the ProgressBar.
         * Defaults to `false`.
         */
        this.indeterminate = false;
        this.displayValue = 0;
        this.previousValue = 0;
        this.localizationChangeSubscription = localization.changes.subscribe(function (_a) {
            var rtl = _a.rtl;
            _this.direction = rtl ? 'rtl' : 'ltr';
        });
    }
    Object.defineProperty(ProgressBarBase.prototype, "isHorizontal", {
        get: function () {
            return this.orientation === 'horizontal';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "isVertical", {
        get: function () {
            return this.orientation === 'vertical';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "disabledClass", {
        get: function () {
            return this.disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "reverseClass", {
        get: function () {
            return this.reverse;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "indeterminateClass", {
        get: function () {
            return this.indeterminate;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "dirAttribute", {
        get: function () {
            return this.direction;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "ariaMinAttribute", {
        get: function () {
            return String(this.min);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "ariaMaxAttribute", {
        get: function () {
            return String(this.max);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "ariaValueAttribute", {
        get: function () {
            return this.indeterminate ? undefined : String(this.displayValue);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "isCompleted", {
        /**
         * @hidden
         */
        get: function () {
            return this.value === this.max;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusWidth", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'horizontal' ? this._progressRatio * 100 : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusHeight", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'vertical' ? this._progressRatio * 100 : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusWrapperWidth", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'horizontal' ? 100 / this._progressRatio : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "statusWrapperHeight", {
        /**
         * @hidden
         */
        get: function () {
            return this.orientation === 'vertical' ? 100 / this._progressRatio : 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarBase.prototype, "_progressRatio", {
        get: function () {
            return calculateRatio(this.min, this.max, this.displayValue);
        },
        enumerable: true,
        configurable: true
    });
    ProgressBarBase.prototype.ngOnChanges = function (changes) {
        var min = extractValueFromChanges(changes, 'min', this.min);
        var max = extractValueFromChanges(changes, 'max', this.max);
        var value = extractValueFromChanges(changes, 'value', this.value);
        if (changes.min || changes.max || changes.value) {
            if (changes.min || changes.max) {
                validateRange(min, max);
            }
            if (changes.value) {
                if (value == null || Number.isNaN(value)) {
                    this.value = min;
                }
                var previousValue = this.displayValue;
                this.displayValue = adjustValueToRange(this.min, this.max, value);
                if (previousValue !== this.displayValue) {
                    this.previousValue = previousValue;
                }
            }
            this.min = min;
            this.max = max;
            this.displayValue = adjustValueToRange(this.min, this.max, value);
        }
    };
    ProgressBarBase.prototype.ngOnDestroy = function () {
        if (this.localizationChangeSubscription) {
            this.localizationChangeSubscription.unsubscribe();
        }
    };
    ProgressBarBase.propDecorators = {
        widgetClasses: [{ type: HostBinding, args: ['class.k-widget',] }, { type: HostBinding, args: ['class.k-progressbar',] }],
        isHorizontal: [{ type: HostBinding, args: ['class.k-progressbar-horizontal',] }],
        isVertical: [{ type: HostBinding, args: ['class.k-progressbar-vertical',] }],
        disabledClass: [{ type: HostBinding, args: ['class.k-state-disabled',] }],
        reverseClass: [{ type: HostBinding, args: ['class.k-progressbar-reverse',] }],
        indeterminateClass: [{ type: HostBinding, args: ['class.k-progressbar-indeterminate',] }],
        dirAttribute: [{ type: HostBinding, args: ['attr.dir',] }],
        roleAttribute: [{ type: HostBinding, args: ['attr.role',] }],
        ariaMinAttribute: [{ type: HostBinding, args: ['attr.aria-valuemin',] }],
        ariaMaxAttribute: [{ type: HostBinding, args: ['attr.aria-valuemax',] }],
        ariaValueAttribute: [{ type: HostBinding, args: ['attr.aria-valuenow',] }],
        max: [{ type: Input }],
        min: [{ type: Input }],
        value: [{ type: Input }],
        orientation: [{ type: Input }],
        disabled: [{ type: Input }],
        reverse: [{ type: Input }],
        indeterminate: [{ type: Input }]
    };
    return ProgressBarBase;
}());

/**
 * Represents the [Kendo UI ProgressBar component for Angular]({% slug overview_progressbar %}).
 *
 * @example
 * ```ts-preview
 * _@Component({
 *    selector: 'my-app',
 *    template: `
 *        <kendo-progressbar [value]="value">
 *        </kendo-progressbar>
 *    `
 * })
 * class AppComponent {
 *     public value = 50;
 * }
 * ```
 */
var ProgressBarComponent = /** @class */ (function (_super) {
    __extends(ProgressBarComponent, _super);
    /**
     * @hidden
     */
    function ProgressBarComponent(localization, zone, renderer) {
        var _this = _super.call(this, localization) || this;
        _this.localization = localization;
        _this.zone = zone;
        _this.renderer = renderer;
        /**
         * Determines whether the status label will be visible.
         * Defaults to `true` - the label will be visible and displayed with the default
         * `LabelSettings` - position: `end` and format: `value`
         */
        _this.label = true;
        /**
         * The animation configuration of the ProgressBar.
         * Defaults to `false`
         */
        _this.animation = false;
        /**
         * The event is fired when the animation indicating the latest value change is finished.
         */
        _this.animationEnd = new EventEmitter();
        return _this;
    }
    Object.defineProperty(ProgressBarComponent.prototype, "showLabel", {
        /**
         * @hidden
         */
        get: function () {
            if (typeof this.label === 'boolean') {
                return this.label;
            }
            else {
                if (this.label && !this.label.hasOwnProperty('visible')) {
                    this.label.visible = true;
                }
                return this.label.visible;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarComponent.prototype, "labelPosition", {
        /**
         * @hidden
         */
        get: function () {
            if (typeof this.label === 'boolean') {
                return 'end';
            }
            else {
                if (this.label && !this.label.hasOwnProperty('position')) {
                    this.label.position = 'end';
                }
                return this.label.position;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarComponent.prototype, "isPositionStart", {
        /**
         * @hidden
         */
        get: function () {
            return this.labelPosition === 'start';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarComponent.prototype, "isPositionCenter", {
        /**
         * @hidden
         */
        get: function () {
            return this.labelPosition === 'center';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarComponent.prototype, "isPositionEnd", {
        /**
         * @hidden
         */
        get: function () {
            return this.labelPosition === 'end';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressBarComponent.prototype, "formattedLabelValue", {
        /**
         * @hidden
         */
        get: function () {
            return formatValue(this.displayValue, this.min, this.max, this.label);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @hidden
     */
    ProgressBarComponent.prototype.ngOnInit = function () {
        if (this.animation && this.previousValue !== this.displayValue) {
            this.startAnimation(0);
        }
    };
    /**
     * @hidden
     */
    ProgressBarComponent.prototype.ngOnChanges = function (changes) {
        _super.prototype.ngOnChanges.call(this, changes);
        if (runAnimation(changes, this.animation, this.previousValue, this.displayValue)) {
            this.startAnimation(this.previousValue);
        }
    };
    /**
     * @hidden
     */
    ProgressBarComponent.prototype.ngOnDestroy = function () {
        if (this.animationStartSubscription) {
            this.animationStartSubscription.unsubscribe();
        }
        if (this.animationFrame) {
            cancelAnimationFrame(this.animationFrame);
        }
    };
    /**
     * @hidden
     */
    ProgressBarComponent.prototype.startAnimation = function (previousValue) {
        this.animate(this.progressStatusElement.nativeElement, previousValue, this.progressStatusWrapperElement.nativeElement);
    };
    Object.defineProperty(ProgressBarComponent.prototype, "animationDuration", {
        /**
         * @hidden
         */
        get: function () {
            if (typeof this.animation === 'boolean') {
                return 400;
            }
            else {
                if (this.animation && !this.animation.hasOwnProperty('duration')) {
                    this.animation.duration = 400;
                }
                return this.animation.duration;
            }
        },
        enumerable: true,
        configurable: true
    });
    ProgressBarComponent.prototype.animate = function (element, previousValue, wrapperElement) {
        var _this = this;
        var isHorizontal = this.orientation === 'horizontal';
        var previousRatio = calculateRatio(this.min, this.max, previousValue);
        var previousStatusWidth = isHorizontal ? previousRatio * 100 : 100;
        var previousStatusHeight = !isHorizontal ? previousRatio * 100 : 100;
        this.zone.runOutsideAngular(function () {
            if (_this.animationFrame) {
                cancelAnimationFrame(_this.animationFrame);
            }
            var property = isHorizontal ? 'width' : 'height';
            var startTime = new Date().getTime();
            var startSize = isHorizontal ? previousStatusWidth : previousStatusHeight;
            var deltaSize = isHorizontal ? _this.statusWidth - previousStatusWidth : _this.statusHeight - previousStatusHeight;
            var duration = _this.animationDuration * Math.abs((deltaSize / 100));
            var animate = function () {
                var elapsed = new Date().getTime() - startTime;
                var position = Math.min(elapsed / duration, 1);
                var width = startSize + deltaSize * position;
                var wrapperWidth = (100 / width) * 100;
                _this.renderer.setStyle(element, property, width + '%');
                _this.renderer.setStyle(wrapperElement, property, wrapperWidth + '%');
                if (position < 1) {
                    requestAnimationFrame(animate);
                }
                else {
                    if (hasObservers(_this.animationEnd)) {
                        _this.animationEnd.emit({
                            from: previousValue,
                            to: _this.displayValue
                        });
                    }
                }
            };
            animate();
        });
    };
    ProgressBarComponent.decorators = [
        { type: Component, args: [{
                    exportAs: 'kendoProgressBar',
                    selector: 'kendo-progressbar',
                    template: "\n        <span class=\"k-progress-status-wrap\"\n            [class.k-progress-start]=\"isPositionStart\"\n            [class.k-progress-center]=\"isPositionCenter\"\n            [class.k-progress-end]=\"isPositionEnd\"\n            [ngStyle]=\"emptyCssStyle\"\n            [ngClass]=\"emptyCssClass\">\n            <span *ngIf=\"showLabel\" class=\"k-progress-status\">{{formattedLabelValue}}</span>\n        </span>\n        <div\n            #progressStatus\n            class=\"k-state-selected\"\n            [class.k-complete]=\"isCompleted\"\n            [ngStyle]=\"progressCssStyle\"\n            [ngClass]=\"progressCssClass\"\n            [style.width.%]=\"statusWidth\"\n            [style.height.%]=\"statusHeight\"\n            >\n            <span\n                #progressStatusWrap\n                class=\"k-progress-status-wrap\"\n                [style.width.%]=\"statusWrapperWidth\"\n                [style.height.%]=\"statusWrapperHeight\"\n                [class.k-progress-start]=\"isPositionStart\"\n                [class.k-progress-center]=\"isPositionCenter\"\n                [class.k-progress-end]=\"isPositionEnd\"\n                >\n                <span *ngIf=\"showLabel\" class=\"k-progress-status\">{{formattedLabelValue}}</span>\n            </span>\n        </div>\n       ",
                    providers: [
                        LocalizationService,
                        {
                            provide: L10N_PREFIX,
                            useValue: 'kendo.progressbar'
                        }
                    ]
                },] },
    ];
    /** @nocollapse */
    ProgressBarComponent.ctorParameters = function () { return [
        { type: LocalizationService },
        { type: NgZone },
        { type: Renderer2 }
    ]; };
    ProgressBarComponent.propDecorators = {
        label: [{ type: Input }],
        progressCssStyle: [{ type: Input }],
        progressCssClass: [{ type: Input }],
        emptyCssStyle: [{ type: Input }],
        emptyCssClass: [{ type: Input }],
        animation: [{ type: Input }],
        animationEnd: [{ type: Output }],
        progressStatusElement: [{ type: ViewChild, args: ['progressStatus',] }],
        progressStatusWrapperElement: [{ type: ViewChild, args: ['progressStatusWrap',] }]
    };
    return ProgressBarComponent;
}(ProgressBarBase));

/**
 * Represents the [Kendo UI ChunkProgressBar component for Angular]({% slug overview_chunkprogressbar %}).
 *
 * @example
 * ```ts-preview
 * _@Component({
 *    selector: 'my-app',
 *    template: `
 *        <kendo-chunkprogressbar [value]="value">
 *        </kendo-chunkprogressbar>
 *    `
 * })
 * class AppComponent {
 *     public value = 40;
 * }
 * ```
 */
var ChunkProgressBarComponent = /** @class */ (function (_super) {
    __extends(ChunkProgressBarComponent, _super);
    /**
     * @hidden
     */
    function ChunkProgressBarComponent(localization) {
        var _this = _super.call(this, localization) || this;
        _this.localization = localization;
        /**
         * Sets the number of chunks into which the ChunkProgressBar will be split.
         * Defaults to `5`.
         */
        _this.chunkCount = 5;
        _this._orientationStyles = {
            width: _this.chunkSizePercentage + "%"
        };
        return _this;
    }
    Object.defineProperty(ChunkProgressBarComponent.prototype, "chunks", {
        /**
         * @hidden
         */
        get: function () {
            var count = this.chunkCount;
            var chunks = Array(count).fill(false);
            var completedChunks = Math.floor(this._progressRatio * count);
            for (var i = 0; i < completedChunks; i++) {
                chunks[i] = true;
            }
            if (reverseChunks(this.orientation, this.reverse)) {
                chunks.reverse();
            }
            return chunks;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ChunkProgressBarComponent.prototype, "chunkSizePercentage", {
        /**
         * @hidden
         */
        get: function () {
            return 100 / this.chunkCount;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ChunkProgressBarComponent.prototype, "orientationStyles", {
        /**
         * @hidden
         */
        get: function () {
            if (this.orientation === 'horizontal') {
                this._orientationStyles.width = this.chunkSizePercentage + "%";
                this._orientationStyles.height = undefined;
            }
            else {
                this._orientationStyles.height = this.chunkSizePercentage + "%";
                this._orientationStyles.width = undefined;
            }
            return this._orientationStyles;
        },
        enumerable: true,
        configurable: true
    });
    ChunkProgressBarComponent.decorators = [
        { type: Component, args: [{
                    exportAs: 'kendoChunkProgressBar',
                    selector: 'kendo-chunkprogressbar',
                    template: "\n        <ul class=\"k-reset\">\n            <li class=\"k-item\" *ngFor=\"let chunk of chunks; let i = index;\"\n                [class.k-first]=\"i === 0\"\n                [class.k-last]=\"i === chunkCount - 1\"\n                [class.k-state-selected]=\"chunk\"\n                [ngClass]=\"chunk ? progressCssClass : emptyCssClass\"\n                [ngStyle]=\"chunk ? progressCssStyle : emptyCssStyle\"\n                [style.width]=\"orientationStyles.width\"\n                [style.height]=\"orientationStyles.height\"\n                >\n            </li>\n        </ul>\n    ",
                    providers: [
                        LocalizationService,
                        {
                            provide: L10N_PREFIX,
                            useValue: 'kendo.chunkprogressbar'
                        }
                    ]
                },] },
    ];
    /** @nocollapse */
    ChunkProgressBarComponent.ctorParameters = function () { return [
        { type: LocalizationService }
    ]; };
    ChunkProgressBarComponent.propDecorators = {
        chunkCount: [{ type: Input }],
        progressCssStyle: [{ type: Input }],
        progressCssClass: [{ type: Input }],
        emptyCssStyle: [{ type: Input }],
        emptyCssClass: [{ type: Input }]
    };
    return ChunkProgressBarComponent;
}(ProgressBarBase));

var COMPONENT_DIRECTIVES = [ProgressBarComponent, ChunkProgressBarComponent];
var MODULES = [CommonModule];
/**
 * @hidden
 */
var ProgressBarModule = /** @class */ (function () {
    function ProgressBarModule() {
    }
    ProgressBarModule.decorators = [
        { type: NgModule, args: [{
                    declarations: COMPONENT_DIRECTIVES,
                    exports: COMPONENT_DIRECTIVES,
                    imports: MODULES
                },] },
    ];
    return ProgressBarModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { ProgressBarBase, ProgressBarComponent, ChunkProgressBarComponent, ProgressBarModule };
