// @progress/kendo-angular-editor/dist/npm/tools/fontfamily/editor-fontfamily-dropdownlist.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2019 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
/**
 * @hidden
 */
var FontFamilyDropDownListComponent = /** @class */ (function () {
    function FontFamilyDropDownListComponent() {
        this.valueChange = new core_1.EventEmitter();
    }
    FontFamilyDropDownListComponent.prototype.onValueChange = function (tag) {
        this.valueChange.emit(tag);
    };
    FontFamilyDropDownListComponent.decorators = [
        { type: core_1.Component, args: [{
                    // tslint:disable-next-line:no-forward-ref
                    selector: 'kendo-editor-fontfamily-dropdownlist',
                    template: "\n        <kendo-dropdownlist\n            #element\n            [defaultItem]=\"defaultItem\"\n            [textField]=\"'text'\"\n            [valueField]=\"'fontName'\"\n            [data]=\"data\"\n            [(value)]=\"value\"\n            [valuePrimitive]=\"true\"\n            [itemDisabled]=\"itemDisabled\"\n            [attr.title]=\"title\"\n            (valueChange)=\"onValueChange($event)\"\n        >\n            <ng-template kendoDropDownListItemTemplate let-dataItem>\n                <span [ngStyle]=\"{ 'font-family': dataItem.fontName }\">\n                    {{ dataItem.text }}\n                </span>\n            </ng-template>\n        </kendo-dropdownlist>\n    "
                },] },
    ];
    FontFamilyDropDownListComponent.propDecorators = {
        data: [{ type: core_1.Input }],
        value: [{ type: core_1.Input }],
        defaultItem: [{ type: core_1.Input }],
        itemDisabled: [{ type: core_1.Input }],
        title: [{ type: core_1.Input }],
        valueChange: [{ type: core_1.Output }],
        element: [{ type: core_1.ViewChild, args: ['element',] }]
    };
    return FontFamilyDropDownListComponent;
}());
exports.FontFamilyDropDownListComponent = FontFamilyDropDownListComponent;
