// @progress/kendo-angular-scheduler/dist/npm/editing/dialogs.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_angular_dialog_1 = require("@progress/kendo-angular-dialog");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var operators_1 = require("rxjs/operators");
var focus_service_1 = require("../navigation/focus.service");
var view_state_service_1 = require("../views/view-state.service");
/**
 * @hidden
 */
var DialogsService = /** @class */ (function () {
    function DialogsService(dialogService, localization, changeDetector, focusService, viewState) {
        this.dialogService = dialogService;
        this.localization = localization;
        this.changeDetector = changeDetector;
        this.focusService = focusService;
        this.viewState = viewState;
        this.isOpen = false;
    }
    DialogsService.prototype.openRemoveConfirmationDialog = function () {
        var _this = this;
        var dialog = this.dialogService.open({
            title: this.textFor('deleteDialogTitle'),
            content: this.textFor('deleteConfirmation'),
            actions: [
                { text: this.textFor('cancel'), value: false },
                { text: this.textFor('destroy'), value: true }
            ],
            appendTo: this.container,
            autoFocusedElement: 'button:nth-child(2)'
        });
        this.isOpen = true;
        this.changeDetector.markForCheck();
        return dialog.result.pipe(operators_1.map(function (result) {
            _this.isOpen = false;
            if (result instanceof kendo_angular_dialog_1.DialogCloseResult) {
                _this.focusService.focus();
                return false;
            }
            _this.viewState.layoutEnd.pipe(operators_1.take(1)).subscribe(function () {
                return _this.focusService.focus();
            });
            var res = result;
            return res.value;
        }));
    };
    DialogsService.prototype.openRecurringConfirmationDialog = function (operation) {
        var _this = this;
        var dialog = this.dialogService.open({
            actions: [
                {
                    text: operation === 0 /* Edit */ ? this.textFor('editOccurrence') : this.textFor('deleteOccurrence'),
                    value: 1 /* Occurrence */
                },
                {
                    text: operation === 0 /* Edit */ ? this.textFor('editSeries') : this.textFor('deleteSeries'),
                    value: 2 /* Series */
                }
            ],
            appendTo: this.container,
            autoFocusedElement: 'button:nth-child(1)',
            content: operation === 0 /* Edit */ ? this.textFor('editRecurringConfirmation') : this.textFor('deleteRecurringConfirmation'),
            title: operation === 0 /* Edit */ ? this.textFor('editRecurringDialogTitle') : this.textFor('deleteRecurringDialogTitle')
        });
        this.isOpen = true;
        this.changeDetector.markForCheck();
        return dialog.result.pipe(operators_1.map(function (result) {
            _this.isOpen = false;
            _this.focusService.focus();
            if (result instanceof kendo_angular_dialog_1.DialogCloseResult) {
                return undefined;
            }
            var res = result;
            return res.value;
        }));
    };
    DialogsService.prototype.textFor = function (key) {
        return this.localization.get(key);
    };
    DialogsService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    DialogsService.ctorParameters = function () { return [
        { type: kendo_angular_dialog_1.DialogService },
        { type: kendo_angular_l10n_1.LocalizationService },
        { type: core_1.ChangeDetectorRef },
        { type: focus_service_1.FocusService },
        { type: view_state_service_1.ViewStateService }
    ]; };
    return DialogsService;
}());
exports.DialogsService = DialogsService;
