// @progress/kendo-angular-editor/dist/npm/index.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2019 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
/**
 * Generated bundle index. Do not edit.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./main"), exports);
var colorpicker_dialog_component_1 = require("./dialogs/colorpicker-dialog.component");
exports.ColorPickerDialogComponent = colorpicker_dialog_component_1.ColorPickerDialogComponent
var file_link_dialog_component_1 = require("./dialogs/file-link-dialog.component");
exports.FileLinkDialogComponent = file_link_dialog_component_1.FileLinkDialogComponent
var font_family_dialog_component_1 = require("./dialogs/font-family-dialog.component");
exports.FontFamilyDialogComponent = font_family_dialog_component_1.FontFamilyDialogComponent
var font_size_dialog_component_1 = require("./dialogs/font-size-dialog.component");
exports.FontSizeDialogComponent = font_size_dialog_component_1.FontSizeDialogComponent
var format_dialog_component_1 = require("./dialogs/format-dialog.component");
exports.FormatDialogComponent = format_dialog_component_1.FormatDialogComponent
var image_dialog_component_1 = require("./dialogs/image-dialog.component");
exports.ImageDialogComponent = image_dialog_component_1.ImageDialogComponent
var insert_table_dialog_component_1 = require("./dialogs/insert-table-dialog.component");
exports.InsertTableDialogComponent = insert_table_dialog_component_1.InsertTableDialogComponent
var source_dialog_component_1 = require("./dialogs/source-dialog.component");
exports.SourceDialogComponent = source_dialog_component_1.SourceDialogComponent
var custom_messages_component_1 = require("./localization/custom-messages.component");
exports.CustomMessagesComponent = custom_messages_component_1.CustomMessagesComponent
var editor_localization_service_1 = require("./localization/editor-localization.service");
exports.EditorLocalizationService = editor_localization_service_1.EditorLocalizationService
var localized_messages_directive_1 = require("./localization/localized-messages.directive");
exports.LocalizedMessagesDirective = localized_messages_directive_1.LocalizedMessagesDirective
var messages_1 = require("./localization/messages");
exports.Messages = messages_1.Messages
var editor_align_center_button_directive_1 = require("./tools/alignment/editor-align-center-button.directive");
exports.EditorAlignCenterButtonDirective = editor_align_center_button_directive_1.EditorAlignCenterButtonDirective
var editor_align_justify_button_directive_1 = require("./tools/alignment/editor-align-justify-button.directive");
exports.EditorAlignJustifyButtonDirective = editor_align_justify_button_directive_1.EditorAlignJustifyButtonDirective
var editor_align_left_button_directive_1 = require("./tools/alignment/editor-align-left-button.directive");
exports.EditorAlignLeftButtonDirective = editor_align_left_button_directive_1.EditorAlignLeftButtonDirective
var editor_align_right_button_directive_1 = require("./tools/alignment/editor-align-right-button.directive");
exports.EditorAlignRightButtonDirective = editor_align_right_button_directive_1.EditorAlignRightButtonDirective
var editor_back_color_directive_1 = require("./tools/colorpicker/editor-back-color.directive");
exports.EditorBackColorDirective = editor_back_color_directive_1.EditorBackColorDirective
var editor_colorpicker_component_1 = require("./tools/colorpicker/editor-colorpicker.component");
exports.EditorColorPickerComponent = editor_colorpicker_component_1.EditorColorPickerComponent
var editor_fore_color_directive_1 = require("./tools/colorpicker/editor-fore-color.directive");
exports.EditorForeColorDirective = editor_fore_color_directive_1.EditorForeColorDirective
var editor_clean_formatting_button_directive_1 = require("./tools/editor-clean-formatting-button.directive");
exports.EditorCleanFormattingButtonDirective = editor_clean_formatting_button_directive_1.EditorCleanFormattingButtonDirective
var editor_fontfamily_dropdownlist_component_1 = require("./tools/fontfamily/editor-fontfamily-dropdownlist.component");
exports.FontFamilyDropDownListComponent = editor_fontfamily_dropdownlist_component_1.FontFamilyDropDownListComponent
var editor_fontfamily_component_1 = require("./tools/fontfamily/editor-fontfamily.component");
exports.EditorFontFamilyComponent = editor_fontfamily_component_1.EditorFontFamilyComponent
var editor_fontsize_dropdownlist_component_1 = require("./tools/fontsize/editor-fontsize-dropdownlist.component");
exports.FontSizeDropDownListComponent = editor_fontsize_dropdownlist_component_1.FontSizeDropDownListComponent
var editor_fontsize_component_1 = require("./tools/fontsize/editor-fontsize.component");
exports.EditorFontSizeComponent = editor_fontsize_component_1.EditorFontSizeComponent
var editor_format_dropdownlist_component_1 = require("./tools/format/editor-format-dropdownlist.component");
exports.FormatDropDownListComponent = editor_format_dropdownlist_component_1.FormatDropDownListComponent
var editor_format_component_1 = require("./tools/format/editor-format.component");
exports.EditorFormatComponent = editor_format_component_1.EditorFormatComponent
var editor_redo_button_directive_1 = require("./tools/history/editor-redo-button.directive");
exports.EditorRedoButtonDirective = editor_redo_button_directive_1.EditorRedoButtonDirective
var editor_undo_button_directive_1 = require("./tools/history/editor-undo-button.directive");
exports.EditorUndoButtonDirective = editor_undo_button_directive_1.EditorUndoButtonDirective
var editor_insert_image_button_directive_1 = require("./tools/image/editor-insert-image-button.directive");
exports.EditorInsertImageButtonDirective = editor_insert_image_button_directive_1.EditorInsertImageButtonDirective
var editor_indent_button_directive_1 = require("./tools/indentation/editor-indent-button.directive");
exports.EditorIndentButtonDirective = editor_indent_button_directive_1.EditorIndentButtonDirective
var editor_outdent_button_directive_1 = require("./tools/indentation/editor-outdent-button.directive");
exports.EditorOutdentButtonDirective = editor_outdent_button_directive_1.EditorOutdentButtonDirective
var editor_create_link_button_directive_1 = require("./tools/link/editor-create-link-button.directive");
exports.EditorCreateLinkButtonDirective = editor_create_link_button_directive_1.EditorCreateLinkButtonDirective
var editor_insert_file_button_directive_1 = require("./tools/link/editor-insert-file-button.directive");
exports.EditorInsertFileButtonDirective = editor_insert_file_button_directive_1.EditorInsertFileButtonDirective
var editor_unlink_button_directive_1 = require("./tools/link/editor-unlink-button.directive");
exports.EditorUnlinkButtonDirective = editor_unlink_button_directive_1.EditorUnlinkButtonDirective
var editor_insert_ordered_list_button_directive_1 = require("./tools/list/editor-insert-ordered-list-button.directive");
exports.EditorInsertOrderedListButtonDirective = editor_insert_ordered_list_button_directive_1.EditorInsertOrderedListButtonDirective
var editor_insert_unordered_list_button_directive_1 = require("./tools/list/editor-insert-unordered-list-button.directive");
exports.EditorInsertUnorderedListButtonDirective = editor_insert_unordered_list_button_directive_1.EditorInsertUnorderedListButtonDirective
var editor_command_base_1 = require("./tools/shared/editor-command-base");
exports.EditorCommandBase = editor_command_base_1.EditorCommandBase
var editor_command_button_1 = require("./tools/shared/editor-command-button");
exports.EditorCommandButton = editor_command_button_1.EditorCommandButton
var editor_command_dialog_1 = require("./tools/shared/editor-command-dialog");
exports.EditorCommandDialog = editor_command_dialog_1.EditorCommandDialog
var editor_view_source_button_directive_1 = require("./tools/source/editor-view-source-button.directive");
exports.EditorViewSourceButtonDirective = editor_view_source_button_directive_1.EditorViewSourceButtonDirective
var editor_add_column_after_button_directive_1 = require("./tools/tables/editor-add-column-after-button.directive");
exports.EditorAddColumnAfterButtonDirective = editor_add_column_after_button_directive_1.EditorAddColumnAfterButtonDirective
var editor_add_column_before_button_directive_1 = require("./tools/tables/editor-add-column-before-button.directive");
exports.EditorAddColumnBeforeButtonDirective = editor_add_column_before_button_directive_1.EditorAddColumnBeforeButtonDirective
var editor_add_row_after_button_directive_1 = require("./tools/tables/editor-add-row-after-button.directive");
exports.EditorAddRowAfterButtonDirective = editor_add_row_after_button_directive_1.EditorAddRowAfterButtonDirective
var editor_add_row_before_button_directive_1 = require("./tools/tables/editor-add-row-before-button.directive");
exports.EditorAddRowBeforeButtonDirective = editor_add_row_before_button_directive_1.EditorAddRowBeforeButtonDirective
var editor_delete_column_button_directive_1 = require("./tools/tables/editor-delete-column-button.directive");
exports.EditorDeleteColumnButtonDirective = editor_delete_column_button_directive_1.EditorDeleteColumnButtonDirective
var editor_delete_row_button_directive_1 = require("./tools/tables/editor-delete-row-button.directive");
exports.EditorDeleteRowButtonDirective = editor_delete_row_button_directive_1.EditorDeleteRowButtonDirective
var editor_delete_table_button_directive_1 = require("./tools/tables/editor-delete-table-button.directive");
exports.EditorDeleteTableButtonDirective = editor_delete_table_button_directive_1.EditorDeleteTableButtonDirective
var editor_insert_table_button_component_1 = require("./tools/tables/editor-insert-table-button.component");
exports.EditorInsertTableButtonComponent = editor_insert_table_button_component_1.EditorInsertTableButtonComponent
var editor_merge_cells_button_directive_1 = require("./tools/tables/editor-merge-cells-button.directive");
exports.EditorMergeCellsButtonDirective = editor_merge_cells_button_directive_1.EditorMergeCellsButtonDirective
var editor_split_cell_button_directive_1 = require("./tools/tables/editor-split-cell-button.directive");
exports.EditorSplitCellButtonDirective = editor_split_cell_button_directive_1.EditorSplitCellButtonDirective
var popup_table_grid_component_1 = require("./tools/tables/popup-table-grid.component");
exports.PopupTableGridComponent = popup_table_grid_component_1.PopupTableGridComponent
var editor_bold_button_directive_1 = require("./tools/typographical-emphasis/editor-bold-button.directive");
exports.EditorBoldButtonDirective = editor_bold_button_directive_1.EditorBoldButtonDirective
var editor_italic_button_directive_1 = require("./tools/typographical-emphasis/editor-italic-button.directive");
exports.EditorItalicButtonDirective = editor_italic_button_directive_1.EditorItalicButtonDirective
var editor_strikethrough_button_directive_1 = require("./tools/typographical-emphasis/editor-strikethrough-button.directive");
exports.EditorStrikethroughButtonDirective = editor_strikethrough_button_directive_1.EditorStrikethroughButtonDirective
var editor_subscript_button_directive_1 = require("./tools/typographical-emphasis/editor-subscript-button.directive");
exports.EditorSubscriptButtonDirective = editor_subscript_button_directive_1.EditorSubscriptButtonDirective
var editor_superscript_button_directive_1 = require("./tools/typographical-emphasis/editor-superscript-button.directive");
exports.EditorSuperscriptButtonDirective = editor_superscript_button_directive_1.EditorSuperscriptButtonDirective
var editor_underline_button_directive_1 = require("./tools/typographical-emphasis/editor-underline-button.directive");
exports.EditorUnderlineButtonDirective = editor_underline_button_directive_1.EditorUnderlineButtonDirective
