// @progress/kendo-angular-editor/dist/es/tools/fontsize/editor-fontsize-dropdownlist.component.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2019 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, ViewChild, Input, ElementRef, Output, EventEmitter } from '@angular/core';
/**
 * @hidden
 */
var FontSizeDropDownListComponent = /** @class */ (function () {
    function FontSizeDropDownListComponent() {
        this.valueChange = new EventEmitter();
    }
    FontSizeDropDownListComponent.prototype.onValueChange = function (size) {
        this.valueChange.emit(size);
    };
    FontSizeDropDownListComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line:no-forward-ref
                    selector: 'kendo-editor-fontsize-dropdownlist',
                    template: "\n        <kendo-dropdownlist\n            #element\n            [defaultItem]=\"defaultItem\"\n            [textField]=\"'text'\"\n            [valueField]=\"'size'\"\n            [data]=\"data\"\n            [(value)]=\"value\"\n            [valuePrimitive]=\"true\"\n            [itemDisabled]=\"itemDisabled\"\n            [attr.title]=\"title\"\n            (valueChange)=\"onValueChange($event)\"\n        >\n        </kendo-dropdownlist>\n    "
                },] },
    ];
    FontSizeDropDownListComponent.propDecorators = {
        data: [{ type: Input }],
        value: [{ type: Input }],
        defaultItem: [{ type: Input }],
        itemDisabled: [{ type: Input }],
        title: [{ type: Input }],
        valueChange: [{ type: Output }],
        element: [{ type: ViewChild, args: ['element',] }]
    };
    return FontSizeDropDownListComponent;
}());
export { FontSizeDropDownListComponent };
