// @progress/kendo-angular-scheduler/dist/es2015/editing/edit-dialog.component.js
import { Component, Input, NgZone, ChangeDetectorRef, ViewChildren, QueryList, ViewChild, ElementRef } from '@angular/core';
import { FormControl } from '@angular/forms';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { EditService } from './edit.service';
import { fromEvent } from 'rxjs';
import { isPresent } from '../common/util';
import { EditDialogTemplateDirective } from './edit-dialog-template.directive';
import { MultipleResourceEditorComponent } from './resource-multiple-editor.component';
import { SingleResourceEditorComponent } from './resource-single-editor.component';
import { Keys } from '@progress/kendo-angular-common';
import { filter, take } from 'rxjs/operators';
import { withModifiers, Modifiers } from '../common/modifiers';
import { FocusService } from '../navigation';
import { ZonedDate, toLocalDate } from '@progress/kendo-date-math';
import { toUTCDateTime, formValueOrDefault } from '../views/utils';
/**
 * @hidden
 */
export class EditDialogComponent {
    constructor(ngZone, editService, localization, changeDetector, element, focusService) {
        this.ngZone = ngZone;
        this.editService = editService;
        this.localization = localization;
        this.changeDetector = changeDetector;
        this.element = element;
        this.focusService = focusService;
        this.resources = [];
        this.timezone = 'Etc/UTC';
        this.setTimeZone = false;
        this.setSeparateStartEndTimeZones = false;
        this.subs = this.editService.changed.subscribe(() => {
            this.ngZone.run(() => { this.onChange(); });
        });
        this.subs.add(fromEvent(this.element.nativeElement, 'keydown')
            .pipe(filter(() => this.isActive))
            .subscribe((e) => {
            if (e.keyCode === Keys.Escape) {
                this.reset();
            }
            if (e.keyCode === Keys.Enter && (withModifiers(e, Modifiers.CtrlKey) || withModifiers(e, Modifiers.MetaKey))) {
                this.onSave(e);
            }
            e.stopPropagation();
        }));
    }
    get isEditingSeries() {
        return this.editMode === 2 /* Series */ && Boolean(this.formGroup.get(this.fields.recurrenceRule));
    }
    get eventTimezone() {
        return formValueOrDefault(this.formGroup, this.fields.startTimezone, this.timezone);
    }
    get endTimezone() {
        return formValueOrDefault(this.formGroup, this.fields.endTimezone, this.eventTimezone);
    }
    onChange() {
        this.changeDetector.markForCheck();
        if (this.editService.hasNewEvent) {
            this.editMode = 2 /* Series */;
            this.formGroup = this.editService.context.formGroup;
            this.isNew = true;
            this.applyLocalTimezone();
        }
        else if (this.editService.isEditing()) {
            const { dataItem, mode } = this.editService.context;
            this.formGroup = this.editService.context.formGroup;
            this.isNew = false;
            this.editedEvent = dataItem;
            this.editMode = isPresent(mode) ? mode : 2 /* Series */;
            this.applyLocalTimezone();
        }
        else {
            this.reset();
            return;
        }
        if (!this.editTemplate) {
            this.addTimeZoneCheckboxesToFormGroup();
            this.subscribeToFormGroupChanges();
        }
        if (isPresent(this.formGroup)) {
            this.recurrenceStart = this.formGroup.get(this.fields.start).value;
        }
        this.ngZone.onStable.pipe(take(1)).subscribe(() => {
            this.titleInput.nativeElement.select();
        });
        this.isActive = true;
    }
    ngOnDestroy() {
        if (this.subs) {
            this.subs.unsubscribe();
        }
    }
    onCancel(e) {
        e.preventDefault();
        this.onClose();
        this.changeDetector.markForCheck();
    }
    onSave(e) {
        e.preventDefault();
        this.applyTimezone();
        this.editService.save();
        this.changeDetector.markForCheck();
    }
    onClose() {
        this.editService.endEdit();
        this.changeDetector.markForCheck();
        this.focusService.focus();
    }
    get hasAllDay() {
        return Boolean(this.formGroup.get(this.fields.isAllDay));
    }
    get hasStartTimeZone() {
        return Boolean(this.formGroup.get(this.fields.startTimezone));
    }
    get isStartTimeZoneVisible() {
        return this.formGroup && this.formGroup.get('startTimezoneCheckbox').value;
    }
    get hasEndTimeZone() {
        return Boolean(this.formGroup.get(this.fields.endTimezone));
    }
    get isEndTimeZoneVisible() {
        return this.formGroup && this.formGroup.get('endTimezoneCheckbox').value;
    }
    getFormValue(field) {
        if (field) {
            return this.formGroup.get(field);
        }
    }
    textFor(key) {
        return this.localization.get(key);
    }
    onResourceClick(resource) {
        const resourceEditors = resource.multiple ? this.multipleResourceEditors : this.singleResourceEditors;
        const currentEditor = resourceEditors.filter((editor) => editor.resource.field === resource.field).pop();
        if (currentEditor) {
            currentEditor.focus();
        }
    }
    reset() {
        this.isActive = false;
        this.setTimeZone = false;
        this.setSeparateStartEndTimeZones = false;
        this.removeTimeZoneCheckboxesFromFormGroup();
        this.focusService.focus();
    }
    addTimeZoneCheckboxesToFormGroup() {
        if (isPresent(this.formGroup)) {
            const startField = this.fields.startTimezone;
            this.formGroup.addControl('startTimezoneCheckbox', new FormControl(this.formGroup.contains(startField) &&
                this.formGroup.get(startField).value));
            const endField = this.fields.endTimezone;
            this.formGroup.addControl('endTimezoneCheckbox', new FormControl(this.formGroup.contains(endField) &&
                this.formGroup.get(endField).value));
        }
    }
    removeTimeZoneCheckboxesFromFormGroup() {
        if (isPresent(this.formGroup)) {
            this.formGroup.removeControl('startTimezoneCheckbox');
            this.formGroup.removeControl('endTimezoneCheckbox');
        }
    }
    subscribeToFormGroupChanges() {
        if (isPresent(this.formGroup)) {
            const fields = this.fields;
            this.formGroup.get('startTimezoneCheckbox').valueChanges.subscribe(isTrue => {
                if (!isTrue) {
                    this.formGroup.get(fields.startTimezone).setValue(null, { emitEvent: false });
                    this.formGroup.get(fields.endTimezone).setValue(null, { emitEvent: false });
                    this.formGroup.get('endTimezoneCheckbox').setValue(false, { emitEvent: false });
                }
            });
            this.formGroup.get('endTimezoneCheckbox').valueChanges.subscribe(isTrue => {
                if (!isTrue) {
                    this.formGroup.get(fields.endTimezone).setValue(null, { emitEvent: false });
                }
            });
            this.formGroup.get(fields.start).valueChanges.subscribe((newStart) => {
                this.recurrenceStart = newStart;
            });
        }
    }
    /**
     * Converts the event dates to "display dates" that look like the original date in its time zone.
     * The result does not represent the same moment in time and must be converted back to local dates.
     */
    applyLocalTimezone() {
        const fields = this.fields;
        const start = this.readDateAsLocal(fields.start, this.eventTimezone);
        const end = this.readDateAsLocal(fields.end, this.endTimezone);
        this.formGroup.get(fields.start).reset(start);
        this.formGroup.get(fields.end).reset(end);
    }
    /**
     * Converts the "display dates" used by the editors back to local dates that represent the true moment in time.
     */
    applyTimezone() {
        const fields = this.fields;
        const start = this.readDateWithTimezone(fields.start, this.eventTimezone);
        const end = this.readDateWithTimezone(fields.end, this.endTimezone);
        this.formGroup.get(fields.start).reset(start);
        this.formGroup.get(fields.end).reset(end);
    }
    readDateWithTimezone(field, timezone) {
        const value = this.formGroup.get(field).value;
        return ZonedDate.fromUTCDate(toUTCDateTime(value), timezone).toLocalDate();
    }
    readDateAsLocal(field, timezone) {
        const value = this.formGroup.get(field).value;
        return toLocalDate(ZonedDate.fromLocalDate(value, timezone).toUTCDate());
    }
}
EditDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'kendo-scheduler-edit-dialog',
                template: `
        <kendo-dialog (close)='onClose()' [minWidth]='700' *ngIf='isActive' title='{{ textFor("editorTitle") }}' class='k-scheduler-edit-dialog'>
            <ng-container *ngIf='!editTemplate'>
                <div class='k-scheduler-edit-form'>
                    <div class='k-edit-form-container'>
                        <form novalidate [formGroup]='formGroup'>
                            <div class='k-edit-label'>
                                <label for='k-event-title'>{{ textFor('editorEventTitle') }}</label>
                            </div>
                            <div class='k-edit-field'>
                                <input #title id='k-event-title' class='k-textbox' placeholder='Title' [formControl]='formGroup.get(fields.title)' />
                            </div>

                            <div class='k-edit-label'>
                                <label (click)="startDateTimePicker.focus()">{{ textFor('editorEventStart') }}</label>
                            </div>
                            <div class='k-edit-field'>
                                <kendo-scheduler-datetime-picker #startDateTimePicker [formControl]='formGroup.get(fields.start)'
                                    [isAllDay]='getFormValue(fields.isAllDay)?.value'>
                                </kendo-scheduler-datetime-picker>
                            </div>

                            <ng-container *ngIf="isStartTimeZoneVisible">
                                <div class='k-edit-label'>
                                    <label (click)="startTzPicker.focus()">{{ textFor('editorEventStartTimeZone') }}</label>
                                </div>

                                <div class='k-edit-field'>
                                    <kendo-timezone-editor #startTzPicker [formControl]='formGroup.get(fields.startTimezone)'></kendo-timezone-editor>
                                </div>
                            </ng-container>

                            <div class='k-edit-label'>
                                <label (click)="endDateTimePicker.focus()">{{ textFor('editorEventEnd') }}</label>
                            </div>
                            <div class='k-edit-field'>
                                <kendo-scheduler-datetime-picker #endDateTimePicker [formControl]='formGroup.get(fields.end)'
                                    [isAllDay]='getFormValue(fields.isAllDay)?.value'>
                                </kendo-scheduler-datetime-picker>
                            </div>

                            <ng-container *ngIf="isEndTimeZoneVisible">
                                <div class='k-edit-label'>
                                    <label (click)="endTzPicker.focus()">{{ textFor('editorEventEndTimeZone') }}</label>
                                </div>

                                <div class='k-edit-field'>
                                    <kendo-timezone-editor #endTzPicker [formControl]='formGroup.get(fields.endTimezone)'></kendo-timezone-editor>
                                </div>
                            </ng-container>

                            <div class='k-edit-field' *ngIf="hasAllDay">
                                <input type='checkbox' id='k-is-allday-chkbox' class='k-checkbox' [formControl]='formGroup.get(fields.isAllDay)' />
                                <label class='k-checkbox-label' for='k-is-allday-chkbox'>{{ textFor('editorEventAllDay') }}</label>
                            </div>

                            <div class='k-edit-field' *ngIf="hasStartTimeZone">
                                <input type='checkbox' id='k-set-timezone' class='k-checkbox'
                                formControlName='startTimezoneCheckbox' />
                                <label class='k-checkbox-label' for='k-set-timezone'>{{ textFor('editorEventTimeZone') }}</label>

                                <ng-container *ngIf="isStartTimeZoneVisible && hasEndTimeZone">
                                    <input type='checkbox' id='k-use-separate' class='k-checkbox' formControlName='endTimezoneCheckbox' />
                                    <label class='k-checkbox-label' for='k-use-separate'>{{ textFor('editorEventSeparateTimeZones') }}</label>
                                </ng-container>
                            </div>

                            <ng-container *ngIf="isEditingSeries">
                                <kendo-recurrence-editor
                                    [formControl]='formGroup.get(fields.recurrenceRule)'
                                    [start]='recurrenceStart'
                                    [timezone]='eventTimezone'
                                ></kendo-recurrence-editor>
                            </ng-container>

                            <ng-container *ngIf='getFormValue(fields.description)'>
                                <div class='k-edit-label'>
                                    <label for='k-event-description'>{{ textFor('editorEventDescription') }}</label>
                                </div>
                                <div class='k-edit-field'>
                                    <textarea id='k-event-description' class='k-textbox' [formControl]='formGroup.get(fields.description)'></textarea>
                                </div>
                            </ng-container>

                            <ng-container *ngFor='let resource of resources'>
                                <ng-container *ngIf='getFormValue(resource.field)'>
                                    <div class='k-edit-label'>
                                        <label (click)="onResourceClick(resource)">
                                            {{ resource.name ? resource.name : resource.field }}
                                        </label>
                                    </div>
                                    <div class='k-edit-field'>
                                        <kendo-multiple-resource-editor
                                            *ngIf='resource.multiple'
                                            [formControl]='formGroup.get(resource.field)'
                                            [resource]='resource'>
                                        </kendo-multiple-resource-editor>
                                        <kendo-single-resource-editor
                                            *ngIf='!resource.multiple'
                                            [formControl]='formGroup.get(resource.field)'
                                            [resource]='resource'>
                                        </kendo-single-resource-editor>
                                    </div>
                                </ng-container>
                            </ng-container>
                        </form>
                    </div>
                </div>
            </ng-container>

            <ng-container *ngIf='editTemplate'>
                <form novalidate [formGroup]='formGroup'>
                    <ng-container [ngTemplateOutlet]='editTemplate.templateRef'
                        [ngTemplateOutletContext]="{
                            $implicit: formGroup,
                            formGroup: formGroup,
                            dataItem: editedEvent,
                            editMode: editMode,
                            isNew: isNew
                        }">
                    </ng-container>
                </form>
            </ng-container>

            <kendo-dialog-actions>
                <button class="k-button" (click)="onCancel($event)">{{ textFor('cancel') }}</button>
                <button class="k-button k-primary" (click)="onSave($event)" [disabled]="!formGroup.valid">{{ textFor('save') }}</button>
            </kendo-dialog-actions>
        </kendo-dialog>
    `
            },] },
];
/** @nocollapse */
EditDialogComponent.ctorParameters = () => [
    { type: NgZone },
    { type: EditService },
    { type: LocalizationService },
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: FocusService }
];
EditDialogComponent.propDecorators = {
    multipleResourceEditors: [{ type: ViewChildren, args: [MultipleResourceEditorComponent,] }],
    singleResourceEditors: [{ type: ViewChildren, args: [SingleResourceEditorComponent,] }],
    titleInput: [{ type: ViewChild, args: ['title',] }],
    resources: [{ type: Input }],
    timezone: [{ type: Input }],
    fields: [{ type: Input }],
    editTemplate: [{ type: Input }]
};
