// @progress/kendo-angular-scheduler/dist/es2015/navigation/focusable.directive.js
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { FocusService } from './focus.service';
/**
 * @hidden
 */
export class FocusableDirective {
    constructor(element, renderer, focusService) {
        this.element = element;
        this.renderer = renderer;
        this.focusService = focusService;
        /**
         * The order of the element with respect to the other focusable elements.
         * If multiple elements share the same value, their relative to each other order follows their position in the component tree.
         */
        this.focusIndex = 0;
        this.toggle(false);
        this.focusService.register(this);
    }
    get visible() {
        return this.element.nativeElement.style.display !== 'none';
    }
    get enabled() {
        return !this.element.nativeElement.disabled;
    }
    ngOnDestroy() {
        this.focusService.unregister(this);
    }
    toggle(active) {
        if (active !== this.active) {
            const index = active ? '0' : '-1';
            this.renderer.setAttribute(this.element.nativeElement, 'tabIndex', index);
            this.active = active;
        }
    }
    canFocus() {
        return this.visible && this.enabled;
    }
    focus() {
        this.element.nativeElement.focus();
    }
    toggleFocus(value) {
        const focusedClass = 'k-state-selected';
        const element = this.element.nativeElement;
        if (value) {
            this.renderer.addClass(element, focusedClass);
        }
        else {
            this.renderer.removeClass(element, focusedClass);
        }
        this.renderer.setAttribute(element, 'aria-selected', value.toString());
    }
}
FocusableDirective.decorators = [
    { type: Directive, args: [{
                selector: '[kendoSchedulerFocusIndex]'
            },] },
];
/** @nocollapse */
FocusableDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: FocusService }
];
FocusableDirective.propDecorators = {
    focusIndex: [{ type: Input, args: ['kendoSchedulerFocusIndex',] }]
};
