// @progress/kendo-angular-scheduler/dist/es/navigation/focus.service.js
import * as tslib_1 from "tslib";
import { ElementRef, Injectable, Renderer2, Optional } from '@angular/core';
import { Subscription } from 'rxjs';
import { DomEventsService } from '../views/common/dom-events.service';
/**
 * @hidden
 */
var FocusService = /** @class */ (function () {
    function FocusService(renderer, wrapper, domEvents) {
        var _this = this;
        this.renderer = renderer;
        this.wrapper = wrapper;
        this.domEvents = domEvents;
        this.items = new Set();
        this.elementMap = new WeakMap();
        this.subs = new Subscription();
        this.subs.add(this.domEvents.focus.subscribe(function (e) { return _this.onFocusIn(e); }));
        this.subs.add(this.domEvents.focusOut.subscribe(function () { return _this.onFocusOut(); }));
    }
    Object.defineProperty(FocusService.prototype, "activeElement", {
        get: function () {
            if (this.activeItem) {
                return this.activeItem.element;
            }
        },
        enumerable: true,
        configurable: true
    });
    FocusService.prototype.ngOnDestroy = function () {
        this.subs.unsubscribe();
    };
    FocusService.prototype.register = function (item) {
        if (!this.activeItem) {
            this.activeItem = item;
            item.toggle(true);
        }
        this.items.add(item);
        this.elementMap.set(item.element.nativeElement, item);
        this.toggleWrapper();
    };
    FocusService.prototype.unregister = function (item) {
        this.items.delete(item);
        this.elementMap.delete(item.element.nativeElement);
        if (item === this.activeItem) {
            this.activateNext();
        }
        this.toggleWrapper();
    };
    FocusService.prototype.focus = function () {
        if (this.activeItem) {
            this.activeItem.focus();
        }
        else {
            this.wrapper.nativeElement.focus();
        }
    };
    FocusService.prototype.focusNext = function (options) {
        var currentItem = this.activeItem;
        this.activateNext(options);
        if (this.activeItem) {
            this.activeItem.focus();
        }
        return this.activeItem !== currentItem;
    };
    FocusService.prototype.activate = function (next) {
        this.items.forEach(function (item) { return item.toggle(item === next); });
        this.activeItem = next;
    };
    FocusService.prototype.activateNext = function (position) {
        this.activeItem = this.findNext(position);
    };
    FocusService.prototype.findNext = function (position) {
        var _a = tslib_1.__assign({ nowrap: false, offset: 1 }, position), offset = _a.offset, nowrap = _a.nowrap;
        var items = Array.from(this.items.values())
            .filter(function (item) { return item.canFocus(); })
            .sort(function (a, b) { return a.focusIndex - b.focusIndex; });
        if (items.length === 0) {
            return null;
        }
        if (!this.activeItem) {
            return nowrap ? null : items[0];
        }
        var index = items.indexOf(this.activeItem);
        var nextIndex = index + offset;
        if (nowrap) {
            nextIndex = Math.max(0, Math.min(items.length - 1, nextIndex));
        }
        else {
            nextIndex = nextIndex % items.length;
            if (nextIndex < 0) {
                nextIndex = items.length - 1;
            }
        }
        return items[nextIndex];
    };
    FocusService.prototype.toggleWrapper = function () {
        if (this.wrapper) {
            this.renderer.setAttribute(this.wrapper.nativeElement, 'tabindex', this.activeItem ? '-1' : '0');
        }
    };
    FocusService.prototype.onFocusIn = function (e) {
        var item = this.elementMap.get(e.target);
        if (!item || item === this.focusedItem) {
            return;
        }
        if (this.focusedItem) {
            this.focusedItem.toggleFocus(false);
        }
        this.activate(item);
        item.toggleFocus(true);
        this.focusedItem = item;
    };
    FocusService.prototype.onFocusOut = function () {
        if (!this.focusedItem) {
            return;
        }
        this.focusedItem.toggleFocus(false);
        this.focusedItem = null;
    };
    FocusService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    FocusService.ctorParameters = function () { return [
        { type: Renderer2, decorators: [{ type: Optional }] },
        { type: ElementRef, decorators: [{ type: Optional }] },
        { type: DomEventsService }
    ]; };
    return FocusService;
}());
export { FocusService };
