// @progress/kendo-angular-scheduler/dist/es/navigation/shortcuts.directive.js
import { ChangeDetectorRef, Directive, NgZone } from '@angular/core';
import { Keys } from '@progress/kendo-angular-common';
import { ZonedDate } from '@progress/kendo-date-math';
import { Subscription } from 'rxjs';
import { FocusService } from './focus.service';
import { Modifiers, noModifiers, withModifiers } from '../common/modifiers';
import { groupResources } from '../common/util';
import { SchedulerComponent } from '../scheduler.component';
import { DomEventsService } from '../views/common/dom-events.service';
import { take } from 'rxjs/operators';
import { ViewStateService } from '../views/view-state.service';
import { hasClasses } from '../common/dom-queries';
import { DialogsService } from '../editing/dialogs.service';
var isContentWrapper = function (element) { return hasClasses(element, 'k-scheduler-content'); };
var ɵ0 = isContentWrapper;
/**
 * @hidden
 */
var ShortcutsDirective = /** @class */ (function () {
    function ShortcutsDirective(scheduler, domEvents, focusService, zone, changeDetector, viewState, dialogsService) {
        var _this = this;
        this.scheduler = scheduler;
        this.domEvents = domEvents;
        this.focusService = focusService;
        this.zone = zone;
        this.changeDetector = changeDetector;
        this.viewState = viewState;
        this.dialogsService = dialogsService;
        this.shortcuts = [{
                match: function (e) { return e.keyCode === Keys.KeyC && noModifiers(e); },
                action: function (e) {
                    var scheduler = _this.scheduler;
                    var hours = new Date().getHours();
                    var selected = scheduler.selectedDate;
                    var start = new Date(selected.getFullYear(), selected.getMonth(), selected.getDate(), hours + 1);
                    var end = new Date(selected.getFullYear(), selected.getMonth(), selected.getDate(), hours + 2);
                    var firstResource;
                    if (scheduler.group) {
                        var resources = scheduler.resources || [];
                        var group = scheduler.group || {};
                        var grouped = groupResources(group, resources);
                        if (grouped.length > 0) {
                            firstResource = grouped[0].data[0];
                        }
                    }
                    scheduler.create.emit({
                        start: ZonedDate.fromLocalDate(start, scheduler.timezone).toLocalDate(),
                        end: ZonedDate.fromLocalDate(end, scheduler.timezone).toLocalDate(),
                        isAllDay: false,
                        resources: [firstResource],
                        originalEvent: e,
                        sender: scheduler
                    });
                }
            }, {
                match: function (e) { return e.keyCode >= Keys.Digit1 && e.keyCode <= Keys.Digit9 && withModifiers(e, Modifiers.AltKey); },
                action: function (e) {
                    var scheduler = _this.scheduler;
                    var viewIndex = e.keyCode - Keys.Digit0 - 1;
                    var views = scheduler.views.toArray();
                    var view = views[viewIndex];
                    if (view) {
                        _this.zone.run(function () {
                            var prevented = scheduler.onNavigationAction({ type: 'view-change', view: view });
                            if (!prevented) {
                                _this.changeDetector.markForCheck();
                                _this.focusWait();
                            }
                        });
                    }
                }
            }, {
                match: function (e) { return e.keyCode === Keys.KeyT && noModifiers(e); },
                action: function () {
                    _this.zone.run(function () {
                        _this.scheduler.onNavigationAction({ type: 'today' });
                        _this.focusWait();
                    });
                }
            }, {
                // tslint:disable-next-line:deprecation
                match: function (e) { return (e.keyCode === Keys.ArrowLeft || e.keyCode === Keys.ArrowRight) && withModifiers(e, Modifiers.ShiftKey); },
                action: function (e) {
                    // tslint:disable-next-line:deprecation
                    var type = e.keyCode === Keys.ArrowLeft ? 'prev' : 'next';
                    _this.zone.run(function () {
                        _this.scheduler.onNavigationAction({ type: type });
                        _this.focusWait();
                    });
                }
            }, {
                // tslint:disable-next-line:deprecation
                match: function (e) { return (e.keyCode === Keys.ArrowUp || e.keyCode === Keys.ArrowLeft) && noModifiers(e) && !isContentWrapper(e.target); },
                action: function (e) {
                    var prevented = _this.scheduler.onNavigationAction({ type: 'focus-prev' });
                    if (!prevented) {
                        _this.focusService.focusNext({ offset: -1 });
                        e.preventDefault();
                    }
                }
            }, {
                // tslint:disable-next-line:deprecation
                match: function (e) { return (e.keyCode === Keys.ArrowDown || e.keyCode === Keys.ArrowRight) && noModifiers(e) && !isContentWrapper(e.target); },
                action: function (e) {
                    var prevented = _this.scheduler.onNavigationAction({ type: 'focus-next' });
                    if (!prevented) {
                        _this.focusService.focusNext({ offset: 1 });
                        e.preventDefault();
                    }
                }
            }];
        this.taskShortcuts = [{
                match: function (e) { return (e.keyCode === Keys.Delete || e.keyCode === Keys.Backspace) && noModifiers(e); },
                action: function (e, event) {
                    _this.viewState.emitEvent('remove', { event: event, dataItem: event.dataItem });
                    e.preventDefault();
                }
            }, {
                match: function (e) { return e.keyCode === Keys.Enter && noModifiers(e); },
                action: function (e, event) {
                    _this.viewState.emitEvent('eventDblClick', { type: 'dblclick', event: event, originalEvent: e });
                    e.preventDefault();
                }
            }];
        this.subs = new Subscription();
        this.subs.add(this.domEvents.keydown.subscribe(function (e) { return _this.onKeydown(e); }));
        this.subs.add(this.scheduler.eventKeydown.subscribe(function (e) { return _this.onEventKeydown(e); }));
    }
    ShortcutsDirective.prototype.onKeydown = function (e) {
        var match = this.shortcuts.find(function (shortcut) { return shortcut.match(e); });
        if (match && !this.dialogsService.isOpen) {
            match.action(e);
        }
    };
    ShortcutsDirective.prototype.onEventKeydown = function (e) {
        var match = this.taskShortcuts.find(function (shortcut) { return shortcut.match(e.originalEvent); });
        if (match && !this.dialogsService.isOpen) {
            match.action(e.originalEvent, e.event);
        }
    };
    ShortcutsDirective.prototype.focusWait = function () {
        var _this = this;
        this.viewState.layoutEnd.pipe(take(1)).subscribe(function () {
            return _this.focusService.focus();
        });
    };
    ShortcutsDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: 'kendo-scheduler'
                },] },
    ];
    /** @nocollapse */
    ShortcutsDirective.ctorParameters = function () { return [
        { type: SchedulerComponent },
        { type: DomEventsService },
        { type: FocusService },
        { type: NgZone },
        { type: ChangeDetectorRef },
        { type: ViewStateService },
        { type: DialogsService }
    ]; };
    return ShortcutsDirective;
}());
export { ShortcutsDirective };
export { ɵ0 };
